lib/tooling/pahole-tool.js

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2018, Compiler Explorer Team
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+"use strict";
+
+const
+    BaseTool = require('./base-tool');
+
+class PaholeTool extends BaseTool {
+    constructor(toolInfo, env) {
+        super(toolInfo, env);
+    }
+
+    runTool(sourcefile, args, compilerExe, outputFilename, option, filters) {
+        if(!filters.binary)
+        {
+            // Pahole needs a compiled file in order to process it
+            var compiledFile = outputFilename.replace("output.s", "output.o");
+            var compilerOptions = [
+                "-c",
```

Comment:
Does this work for every C++ compiler?

Reply:
It works on linux only, for every compiler that produces elf files. Object files, executables, as long as it's compiled

Reply:
I suspect we'll need an override point for the compiler here; somehow working out if it's even supported by the underlying compiler and if so which options to pass it. This is a great start though.

I'll be honest I've not looked at the tooling framework too much (whoops), so I see there's no `compiler` object passed in here we could interrogate, which makes this tricky. To start with this is cool though! We may have to adapt the tool framework (in a followup change). I'm not against merging this to start with.

Reply:
I am not familiar with the concept and the syntax for the promise in javascript, does this looks acceptable ?